src/main/java/org/opentripplanner/routing/vehicle_parking/VehicleParkingService.java

```diff
@@ -1,22 +1,66 @@
 package org.opentripplanner.routing.vehicle_parking;
 
+import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.ImmutableListMultimap;
+import com.google.common.collect.Multimap;
 import java.io.Serializable;
+import java.util.Collection;
 import java.util.HashSet;
 import java.util.Set;
 import java.util.stream.Stream;
 
+/**
+ * Service that holds all the {@link VehicleParking} instances and an index for fetching parking
+ * locations within a {@link VehicleParkingGroup}. This class is thread-safe because the collections
+ * held here are immutable and only updated in atomic operations that replace the existing
+ * collection with a new copy.
+ *
+ * <P>THIS CLASS IS THREAD-SAFE<p>
+ */
 public class VehicleParkingService implements Serializable {
 
   private static final long serialVersionUID = 1L;
 
-  private final Set<VehicleParking> vehicleParkings = new HashSet<>();
+  /**
+   * To ensure that his is thread-safe, the set stored here should always be immutable.
+   */
+  private Set<VehicleParking> vehicleParkings = Set.of();
 
-  public void addVehicleParking(VehicleParking vehicleParking) {
-    vehicleParkings.add(vehicleParking);
-  }
+  /**
+   * To ensure that his is thread-safe, {@link ImmutableListMultimap} is used.
+   */
+  private ImmutableListMultimap<VehicleParkingGroup, VehicleParking> vehicleParkingGroups = ImmutableListMultimap.of();
+
+  /**
+   * Does atomic update of {@link VehicleParking} and index of {@link VehicleParkingGroup} in this
+   * service by replacing the existing with a new copy that includes old ones that were not removed
+   * in the update and the new ones that were added in the update.
+   */
+  public void updateVehicleParking(
```

Comment:
This method should be synchronized, if calling simultaneously from multiple updaters?

Reply:
For some reason I was under the impression that there was a thread pool of one thread reserved for the updaters but now I realize that it wouldn't work. It's the scheduler that runs on a single thread.

Reply:
You are correct. `GraphUpdaterManager` has one `ExecutorService` which runs all writes in a single thread.

Reply:
So does that mean the method needs or doesn't need to be synchronized?

Reply:
doesn't need to be